plugin/cql/src/main/java/org/opencds/cqf/ruler/cql/KnowledgeArtifactProcessor.java

```diff
@@ -574,6 +584,229 @@ private Optional<MetadataResource> checkIfReferenceInList(RelatedArtifact artifa
+		){
+		if (resource != null) {
+			KnowledgeArtifactAdapter<MetadataResource> adapter = new KnowledgeArtifactAdapter<MetadataResource>(resource);
+			findUnsupportedCapability(resource, capability);
+			processCanonicals(resource, canonicalVersion, checkCanonicalVersion, forceCanonicalVersion);
+			boolean entryExists = bundle.getEntry().stream()
+				.map(e -> (MetadataResource)e.getResource())
+				.filter(mr -> mr.getUrl() != null && mr.getVersion() != null)
+				.anyMatch(mr -> mr.getUrl().equals(resource.getUrl()) && mr.getVersion().equals(resource.getVersion()));
+			if (!entryExists) {
+				BundleEntryComponent entry = createEntry(resource);
+				// TODO: remove history from request.url
+				entry.getRequest().setMethod(HTTPVerb.POST);
+				entry.getRequest().setIfNoneExist("url="+resource.getUrl()+"&version="+resource.getVersion());
+				bundle.addEntry(entry);
+			}
+			List<RelatedArtifact> components = adapter.getComponents();
+			List<RelatedArtifact> dependencies = adapter.getDependencies();
+			Stream.concat(components.stream(), dependencies.stream())
+				.map(ra -> searchResourceByUrl(ra.getResource(), fhirDal))
+				.map(searchBundle -> searchBundle.getEntry().stream().findFirst().orElseGet(()-> new BundleEntryComponent()).getResource())
+				.forEach(component -> recursivePackage((MetadataResource)component, bundle, fhirDal, capability, include, canonicalVersion, checkCanonicalVersion, forceCanonicalVersion));
+		}
+	}
+	private Optional<String> findVersionInListMatchingResource(List<CanonicalType> list, MetadataResource resource){
+		return list.stream()
+					.filter((canonical) -> Canonicals.getUrl(canonical).equals(resource.getUrl()))
+					.map((canonical) -> Canonicals.getVersion(canonical))
+					.findAny();
+	}
+	private void findUnsupportedCapability(MetadataResource resource, List<String> capability) throws PreconditionFailedException{
+		if (capability != null) {
+			List<Extension> knowledgeCapabilityExtension = resource.getExtension().stream()
+			.filter(ext -> ext.getUrl().contains("cqf-knowledgeCapability"))
+			.collect(Collectors.toList());
+			if (knowledgeCapabilityExtension.isEmpty()) {
+				// consider resource unsupported
+				// if it's knowledgeCapability is
+				// undefined
+				throw new PreconditionFailedException(String.format("Resource with url: '%s' does not specify capability.", resource.getUrl()));
+			}
+			knowledgeCapabilityExtension.stream()
+				.filter(ext -> !capability.contains(((CodeType) ext.getValue()).getValue()))
+				.findAny()
+				.ifPresent((ext) -> {
+					throw new PreconditionFailedException(String.format("Resource with url: '%s' is not one of '%s'.",
+					resource.getUrl(),
+					String.join(", ", capability)));
+				});
+		}
+	}
+	private void processCanonicals(MetadataResource resource, List<CanonicalType> canonicalVersion,  List<CanonicalType> checkCanonicalVersion,  List<CanonicalType> forceCanonicalVersion) throws UnprocessableEntityException {
```

Comment:
I'm assuming for $package there is no need to update the relatedArtifact references even though we might modify versions in the output. 

For example if there is a reference to a leaf resource in the root resource `relatedArtifact`, and the version of that leaf resource is changed via `forceCanonicalVersion`, I'm assuming that we don't need to propagate that change to the root resource relatedArtifact reference.

Reply:
That's correct. It is valid to have unversionsed references and any versions specified in context of the $package operation shouldn't be persisted beyond the scope of operation/call.